keps/sig-node/127-user-namespaces/README.md

```diff
@@ -0,0 +1,423 @@
+some workload, a clear error will be shown.
+
+PLEASE note that only phase 1 is targeted for alpha. Also note that the missing
+details (CRI changes, changes needed in container runtimes, etc.) will be added
+in a follow-up PRs.
+
+Please note the last sub-section here is a table with the summary of the changes
+proposed on each phase. That table is not updated (it is from the initial
+proposal, doesn't have all the feedback and adjustments we discussed) but can
+still be useful as a general overview.
+
+
+#### Phase 1: pods "without" volumes
+
+This phase makes pods "without" volume work with user namespaces. This is
+activated via the bool `pod.spec.HostUsers` and can only be set set to `false`
+on pods which use either no volumes or only volumes of the following types:
+
+ - configmap
+ - secret
+ - downwardAPI
+ - emptyDir
+ - projected
+
+This list of volumes was chosen as they can't be used to share files with other
+pods.
+
+The mapping length will be 65535, mapping the range 0-65534 to the pod. This wide
+range makes sure most workloads will work fine.
+
+The mapping will be chosen by the kubelet, using a simple algorithm to give
+different pods in this category ("without" volumes) a non-overlapping mapping.
+Giving non-overlapping mappings generates the best isolation for pods.
+
+Furthermore, the node UID space of 2^32 can hold up to 2^16 pods each with a
+mapping length of 65535 (2^16-1) top. This imposes a limit of 65k pods per node,
+but that is not an issue we will hit in practice for a long time, if ever (today
+we run 110 pods per node by default).
+
+#### Phase 2: pods with volumes
+
+This phase makes user namespaces work in pods with volumes too. This is
+activated via the bool `pod.spec.HostUsers` too and pods fall into this mode if
+some other volume type than the ones listed for phase 1 is used. IOW, when phase
+2 is implemented, pods that use volume types not supported in phase 1, fall into
+the phase 2.
+
+All pods in this mode will use _the same_ mapping, chosen by the kubelet, with a
+length 65535, and mapping the range 0-65534 too. IOW, each pod will have its own user
+namespace, but they will map to _the same_ UIDs/GIDs in the host.
+
+Using the same mapping allows for pods to share files and mitigates all the
+listed vulnerabilities (as the host is protected from the container). It is also
+a trivial next-step to take, given that we have phase 1 implemented: just return
+the same mapping if the pod has other volumes.
+
+#### Phase 3: TBD
+
+#### Unresolved
+
+Here is a list of considerations raised in PRs discussion that hasn't yet
+settle. This list is not exhaustive, we are just trying to put the things that
+we don't want to forget or want to highlight it. Some things that are obvious we
+need to tackle are not listed. Let us know if you think it is important to add
+something else to this list:
+
+- We will start with mappings of 64K. Tim Hockin, however, has expressed
```

Comment:
Yes, exactly, it is a guessing game (people argue the chances you have to guess, though :)). I like that alternative quite a lot, thanks!

I'll add a sentence mentioning the alternative, so we don't forget when we discuss again how to address this item. For now, if it sounds good, I'll just add your alternative there so we can merge the KEP and step by step resolve the unresolved steps in future PRs :)

Reply:
Added the sentence with a reference to your comment. Thanks!

Reply:
We have the option of allowing Pods to express what IDs they use.  In so doing, we provide a way for the Pod to bypass the guessing heuristic.

IMO we MUST have such a capability if the size of the allocation to the Pod ever goes lower than 64K, whether as a fixed (hardcoded), dynamic (in response to maxPodsPerNode) or tunable (by cluster admin) value.

Reply:
choosing the IDs has a visible effect inside the container.  If idmapped mounts are supported, choosing the user namespace mappings don't have any visible effect for the container payload (unless you look into /proc/*/[ug]id_map`).

In what cases would you like to override these mappings?  If we allow overriding the mappings, then there is the risk they could conflict with another pod already running, and I think that should be a privileged operation.

Reply:
@giuseppe +1

In any case, when we resolve that item, this thread is already linked and we will take this into account. IMHO, this should be enough for now.

Reply:
@giuseppe For clarification, I mean that the Pod should be able to express what *container* UIDs it requires, so that Kubelet can ensure that all those IDs get mapped in the user namespace.  I do not think that a Pod should be able to request particular *host* UIDs.  And I agree that such a capability, if implemented, must be a privileged operation.  But I don't think it's necessary.

Reply:
I'm not convinced that we MUST have the ability for the pod to express which container IDs are needed if we go bellow 64k mappings. I'm not sure it is worth having shorter than 64k mappings if we need to request applications to specify which IDs they need.

But let's discuss this when we have to make a decision and have concrete trade-offs for different options. Right now is quite a theoretical issue of what we might do if we need to do X or Y.

Reply:
@frasertweedale also, just curious: why do you think shorter than 64k mappings might require apps to request specific IDs? I mean, I agree that 64k mappings will make things "just work", but I'm curious about your reasons :)

Reply:
@rata simply this: no heuristic about what IDs are required can be guaranteed to cover all use cases.

A more concrete example: most linux distros use UID 65534 as `nobody`, and unmapped UIDs in the user-ns will map to that ID.  Therefore the default heuristic will probably have to include 65534 in the mapping.  But historically not all distros used UID 65534 for `nobody`.

Less concretely, it is probably the case that all kinds of exotic workloads use who-knows-what UIDs.  If it is not possible that a given heuristic can cover all use cases, then it follow that a way to work around that heuristic is needed.

It's not necessary to decide on this for phase 1, and it's not a hill to die on, but I do think it's important.  I will not be surprised when someone's workload can't run if we ship this feature without a way to bypass the "guessing game" (which you previously agreed it is).

Reply:
@rata as to whether it's not worth it to allow <64k mappings if it entails the described complexity, I'm comfortable with that outcome.

Reply:
In kubeadm we implemented rootless mode for the managed static pods by allocating available uid/gid from a node.
https://github.com/kubernetes/kubeadm/issues/2473
(Alpha, has KEP link)
We might want to see user namespace support instead.

On the topic of Linux standardization, we found that the only guaranteed asset is the existence of /etc/users and /etc/groups files and their formatting. There were differences across distros on what are system reserved IDs and how the common CLI tools worked. Thus we wrote our own parser for the user/group files.

Reply:
@frasertweedale Thanks for your reasons. Which distros used a different UID for nobody? Do you know when that was the case?

@neolit123 thanks! I hope this KEP lands and you can use it in kubeadm "soon" :)

Reply:
@rata e.g. Fedora, where `nobody` used to be uid 99, until f28 in mid-2018.  See https://fedoraproject.org/wiki/Changes/RenameNobodyUser for details.

Reply:
@frasertweedale cool, thanks!

Reply:
Marking this as solved, as I already added an item about this in the unresolved list with a link to the comment starting this thread.

Thanks!